CrossPare/src/main/java/de/ugoe/cs/cpdp/dataselection/OutOfSampleBootstrap.java

```diff
@@ -0,0 +1,152 @@
+     *
+     * @see de.ugoe.cs.cpdp.dataselection.IPointWiseDataselectionStrategy#apply(de.ugoe.cs.cpdp.versions.SoftwareVersion,
+     * de.ugoe.cs.cpdp.versions.SoftwareVersion)
+     */
+    @Override
+    public SoftwareVersion apply(SoftwareVersion testversion, SoftwareVersion trainversion) {
+        WekaUtils.makeClassBinary(testversion.getInstances());
+        WekaUtils.makeClassBinary(trainversion.getInstances());
+
+        SoftwareVersion testversioncopy = new SoftwareVersion(testversion);
+        Instances testdata = testversion.getInstances();
+        Instances testBugMatrix = testversion.getBugMatrix();
+        List<Double> testEfforts = testversion.getEfforts();
+        List<Double> testNumBugs = testversion.getNumBugs();
+
+        Instances traindata = trainversion.getInstances();
+        Instances trainBugMatrix = trainversion.getBugMatrix();
+        List<Double> trainEfforts = trainversion.getEfforts();
+        List<Double> trainNumBugs = trainversion.getNumBugs();
+
+        boolean takeNewBootstrapSample = true;
+        final int abortLimit = 10000;
+        int abortCounter = 0;
+        final int datasetSize = testdata.size();
+        Random rand = new Random();
+
+        while (takeNewBootstrapSample){
+            List<Integer> bootstrapSample = new ArrayList<>();
+            for (int i = 0; i < datasetSize; i++) {
+                bootstrapSample.add(rand.nextInt(datasetSize));
+            }
+            traindata.clear();
+            if (trainBugMatrix != null) {
+                trainBugMatrix.clear();
+            }
+            if (trainEfforts != null) {
+                trainEfforts.clear();
+            }
+            if (trainNumBugs != null) {
+                trainNumBugs.clear();
+            }
+            for (int i = 0; i < datasetSize; i++) {
+                int index = bootstrapSample.get(i);
+                traindata.add((Instance) testversioncopy.getInstances().get(index).copy());
+                if (trainBugMatrix != null) {
+                    trainBugMatrix.add((Instance) testversioncopy.getBugMatrix().get(index).copy());
+                }
+                if (trainEfforts != null) {
+                    trainEfforts.add(testversioncopy.getEfforts().get(index));
+                }
+                if (trainNumBugs != null) {
+                    trainNumBugs.add(testversioncopy.getNumBugs().get(index));
+                }
+            }
+            if (traindata.attributeStats(traindata.classIndex()).nominalCounts[1] >= this.minDefects){
+                takeNewBootstrapSample = false;
+            }
+            else {
+                abortCounter += 1;
+                if (abortCounter >= abortLimit){
+                    throw new RuntimeException(
+                            String.format("Error while taking a bootstrap sample of %s: After %d iterations, " +
+                                            "no bootstrap sample containing a defect was created.",
```

Comment:
this message is not accurage. there may be defects, but not more than minDefects